File: docs/guides/dart/lib/saveObjectsChunks.dart
import 'dart:convert';
import 'dart:io';

import 'package:algolia_client_search/algolia_client_search.dart';

void saveObjectsChunks() async {
  final path = '/tmp/records.json';
  final json = File(path).readAsStringSync();
  List<Map<String, dynamic>> records;
  try {
    records = List<Map<String, dynamic>>.from(jsonDecode(json));
  } catch (e) {
    throw Exception('Failed to read file at $path: $e');
  }

  final client =
      SearchClient(appId: 'ALGOLIA_APPLICATION_ID', apiKey: 'ALGOLIA_API_KEY');

  final chunkSize = 10000;

  for (var i = 0; i < records.length; i += chunkSize) {
    final chunk = records.sublist(
        i, (i + chunkSize > records.length) ? records.length : i + chunkSize);
    final batchParams = BatchWriteParams(
        requests: chunk
            .map((record) => BatchRequest(
                  action: Action.addObject,
                  body: record,
                ))
            .toList());
    await client.batch(
      indexName: "<YOUR_INDEX_NAME>",
      batchWriteParams: batchParams,
    );
  }
}

File: docs/guides/scala/src/main/scala/saveObjectsChunks.scala
import scala.concurrent.Await
import scala.concurrent.duration.Duration
import scala.concurrent.Future
import scala.concurrent.ExecutionContext.Implicits.global
import scala.io.Source
import scala.util.Using

import org.json4s.native.JsonMethods
import org.json4s.jvalue2extractable

import algoliasearch.api.SearchClient
import algoliasearch.config.*
import algoliasearch.extension.SearchClientExtensions
import algoliasearch.search.JsonSupport
import algoliasearch.config.RequestOptions

def saveObjectsChunks(): Future[Unit] = {
  implicit val formats: org.json4s.Formats = JsonSupport.format

  val path = "/tmp/records.json"
  val result = Using(Source.fromFile(path))(_.mkString).getOrElse {
    throw new RuntimeException(s"Failed to read file at $path")
  }
  val records = JsonMethods.parse(result).extract[Seq[Map[String, Any]]]

  val client = SearchClient(appId = "ALGOLIA_APPLICATION_ID", apiKey = "ALGOLIA_API_KEY")

  records.grouped(10000).foreach { chunk =>
    try {
      Await.result(
        client.saveObjects(
          indexName = "<YOUR_INDEX_NAME>",
          objects = chunk
        ),
        Duration(5, "sec")
      )
    } catch {
      case e: Exception => println(e)
    }
  }

  Future.unit
}
